rust/mdb_shard/src/shard_file_manager.rs

```diff
@@ -210,17 +185,19 @@ impl ShardFileManager {
             }
         }
 
-        if let Some((si, _)) = self
-            .query_all_shards(&mut |si, f, file_name| {
-                debug!("Querying for hash {:?} in {file_name:?}.", &query_hashes[0]);
-                si.chunk_hash_dedup_query(f, query_hashes)
-            })
-            .await?
-        {
-            Ok(Some(si))
-        } else {
-            Ok(None)
+        for (si, is_permanent) in self.shard_file_lookup.read().await.values() {
+            debug!("Querying for hash {:?} in {:?}.", &query_hashes[0], si.path);
+            if let Some((count, fdse)) = si.chunk_hash_dedup_query(query_hashes)? {
+                if *is_permanent {
```

Comment:
This doesn't track references to shards in the session directory. Is that the design choice? If we don't track I don't see the point to merge shards in the order of mtime?

Reply:
That is correct.  Because of merging, we don't know the hashes for those shards, so we do this part when we're processing the session directory.  See  https://github.com/xetdata/xet-core/pull/66#issue-1816374906 -- this explains why it must be done in order.